File: Gramaticas/Ejercicio1/Python/ejercicio1.py
# Este programa revisa si las cadenas que están en el archivo g1.txt
# son números capicúas formados únicamente por los símbolos 0 y 1.

def es_ejercicio1(cadena):
    # Recorremos cada caracter de la cadena y verificamos que sea 0 o 1
    for caracter in cadena:
        if caracter != "0" and caracter != "1":
            # Si aparece otro símbolo, ya no pertenece al lenguaje
            return False
    
    # Revisamos si la cadena es igual de adelante hacia atrás y de atrás hacia adelante
    if cadena == cadena[::-1]:
        return True
    else:
        return False

def procesar(archivo):
    # Abrimos el archivo de prueba
    with open(archivo, "r") as f:
        for linea in f:
            # Quitamos espacios en blanco y saltos de línea
            cad = linea.strip()
            # Verificamos con la función
            if es_ejercicio1(cad):
                print(cad, "=> ACEPTA")
            else:
                print(cad, "=> NO ACEPTA")

if __name__ == "__main__":
    procesar("ejercicio1.txt")

File: Gramaticas/Ejercicio1/C/ejercicio1.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

// Función que revisa si una cadena es capicúa y solo tiene 0 y 1
bool es_ejercicio1(char *s) {
    int n = strlen(s);

    // Recorremos la cadena para verificar que solo tenga 0 o 1
    for (int i = 0; i < n; i++) {
        if (s[i] != '0' && s[i] != '1') {
            return false;
        }
    }

    // Ahora revisamos si es capicúa
    for (int i = 0; i < n/2; i++) {
        if (s[i] != s[n - 1 - i]) {
            return false;
        }
    }

    return true;
}

int main() {
    FILE *f = fopen("ejercicio1.txt", "r");
    if (f == NULL) {
        printf("No se pudo abrir el archivo\n");
        return 1;
    }

    char linea[100];
    while (fgets(linea, sizeof(linea), f)) {
        linea[strcspn(linea, "\n")] = 0; // Quitar salto
        if (es_ejercicio1(linea)) {
            printf("%s => ACEPTA\n", linea);
        } else {
            printf("%s => NO ACEPTA\n", linea);
        }
    }

    fclose(f);
    return 0;
}
